api/server/httputils/form.go

```diff
@@ -109,3 +115,34 @@ func ArchiveFormValues(r *http.Request, vars map[string]string) (ArchiveOptions,
 	}
 	return ArchiveOptions{name, path}, nil
 }
+
+// ParsePlatform parses the OCI platform from a form
+func ParsePlatform(form url.Values) (*ocispec.Platform, error) {
+	var p ocispec.Platform
+
+	platformJson := form.Get("platform")
+	if platformJson == "" {
+		return nil, nil
+	}
+
+	if err := json.Unmarshal([]byte(platformJson), &p); err != nil {
+		return nil, errdefs.InvalidParameter(errors.Wrap(err, "failed to parse platform"))
+	}
+
+	hasAnyOptional := (p.Variant != "" || p.OSVersion != "" || len(p.OSFeatures) > 0)
+
+	if p.OS == "" && p.Architecture == "" && hasAnyOptional {
+		return nil, errdefs.InvalidParameter(errors.New("optional platform fields provided, but OS and Architecture are missing"))
+	}
+
+	if p.OS == "" && p.Architecture == "" {
+		// No platform specified
+		return nil, nil
+	}
+
+	if (p.OS != "") != (p.Architecture != "") {
```

Comment:
We already check for both to be empty (which we allow), perhaps it's easier to read as;

```suggestion
	if p.OS == "" || p.Architecture == "" {
```